### PR TITLE
Add possibility to finalize the checkout with paid transactions

### PR DESCRIPTION
**Pull request by korycins**
I want to merge this change because it covers https://github.com/saleor/saleor/issues/12303

With this PR we allow using checkoutComplete mutation when the checkout is paid with transactions.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/checkout/complete_checkout.py [hunk 1]
```diff
@@ -711,6 +699,64 @@ def _prepare_checkout(
         checkout.save(update_fields=to_update)
 
 
+def _prepare_checkout_with_transactions(
+    manager: "PluginsManager",
+    checkout_info: "CheckoutInfo",
+    lines: Iterable["CheckoutLineInfo"],
+    discounts,
+    tracking_code,
+    redirect_url,
```
Comment (IKarbowiak) [inline saleor/checkout/complete_checkout.py:708]: Could you add typing to all arguments?

File: saleor/checkout/complete_checkout.py [hunk 2]
```diff
@@ -711,6 +699,64 @@ def _prepare_checkout(
...
+
+
+def _prepare_checkout_with_payment(
+    manager: "PluginsManager",
+    checkout_info: "CheckoutInfo",
+    lines: Iterable["CheckoutLineInfo"],
+    discounts,
+    tracking_code,
+    redirect_url,
+    payment,
```
Comment (IKarbowiak) [inline saleor/checkout/complete_checkout.py:739]: Same here 👮‍♀️

File: saleor/checkout/tests/test_checkout_complete.py [hunk 3]
```diff
@@ -1724,3 +1719,127 @@ def test_complete_checkout_invalid_shipping_method(
...
+        lines=lines,
+        payment_data={},
+        store_source=False,
+        discounts=None,
+        user=customer_user,
+        app=None,
+    )
+
+    # then
+    assert mocked_flow.called
```
Comment (IKarbowiak) [inline saleor/checkout/tests/test_checkout_complete.py:1805]: Same here

File: saleor/checkout/tests/test_checkout_complete.py [hunk 4]
```diff
@@ -1724,3 +1719,127 @@ def test_complete_checkout_invalid_shipping_method(
...
```
Comment (IKarbowiak) [inline saleor/checkout/tests/test_checkout_complete.py:1845]: And here 😄

File: saleor/graphql/checkout/tests/mutations/test_checkout_complete_with_transactions.py [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+    redirect_url = "https://www.example.com"
+    variables = {"id": to_global_id_or_none(checkout), "redirectUrl": redirect_url}
+
+    # when
+    response = user_api_client.post_graphql(MUTATION_CHECKOUT_COMPLETE, variables)
+
+    # then
+    content = get_graphql_content(response)
+    data = content["data"]["checkoutComplete"]
+    assert data["errors"]
```
Comment (IKarbowiak) [inline saleor/graphql/checkout/tests/mutations/test_checkout_complete_with_transactions.py:118]: We should check something more here: the number of errors, `code`, and `field`.

File: saleor/checkout/complete_checkout.py [hunk 6]
```diff
@@ -1163,6 +1205,22 @@ def _create_order_from_checkout(
         prices_entered_with_tax=prices_entered_with_tax,
     )
 
+    # update undiscounted order total
+    undiscounted_total = (
+        sum(
+            [line.line.undiscounted_total_price for line in order_lines_info],
+            start=zero_taxed_money(taxed_total.currency),
```
Comment (maarcingebala) [inline saleor/checkout/complete_checkout.py:1212]: I'm wondering why we're changing this. Maybe it's correct, but I'm just checking if it won't cause unexpected calculation changes.

Comment (korycins) [reply saleor/checkout/complete_checkout.py:1212]: @maarcingebala I've found out, that we fixed checkoutComplete mutation, but we didn't change the part related to the orderCreateFromCheckout. So I copy&paste the fix from the checkoutComplete.

File: saleor/checkout/complete_checkout.py [hunk 7]
```diff
@@ -1272,6 +1330,95 @@ def create_order_from_checkout(
...
+    tracking_code: Optional[str] = None,
+    redirect_url: Optional[str] = None,
+    metadata_list: Optional[List] = None,
+    private_metadata_list: Optional[List] = None,
+) -> Tuple[Optional[Order], bool, dict]:
+    transactions = checkout_info.checkout.payment_transactions.all()
+    fetch_checkout_data(checkout_info, manager, lines, discounts=discounts)
+    checkout_is_zero = checkout_info.checkout.total.gross.amount == Decimal(0)
+    if transactions or (
+        checkout_is_zero
```
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/checkout/complete_checkout.py:1351]: @maarcingebala  When checkout is zero, you don't need any transaction to cover the checkout total. We check if checkout is zero, and we also check what flow for marking an order as paid is used.
Do you want me to provide here comment in the code?

[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/checkout/complete_checkout.py:1351]: I think some small explanation would be fine.

**Review by maarcingebala — APPROVED** [review]

File: saleor/checkout/complete_checkout.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/checkout/complete_checkout.py:1356]: Added

**Review by kadewu — APPROVED** [review]

**Pull request merged**